src/backend/distributed/master/master_repair_shards.c

```diff
@@ -457,12 +423,89 @@ ReplicateColocatedShardPlacement(int64 shardId, char *sourceNodeName,
 		EnsureReferenceTablesExistOnAllNodes();
 	}
 
+	CopyShardTables(colocatedShardList, sourceNodeName, sourceNodePort,
+					targetNodeName, targetNodePort);
+
 	/*
-	 * CopyColocatedShardPlacement function copies given shard with its co-located
-	 * shards.
+	 * Finally insert the placements to pg_dist_placement and sync it to the
+	 * metadata workers.
 	 */
-	CopyColocatedShardPlacement(shardId, sourceNodeName, sourceNodePort,
-								targetNodeName, targetNodePort);
+	foreach_ptr(colocatedShard, colocatedShardList)
+	{
+		uint64 colocatedShardId = colocatedShard->shardId;
+		uint32 groupId = GroupForNode(targetNodeName, targetNodePort);
+		uint64 placementId = GetNextPlacementId();
+
+		InsertShardPlacementRow(colocatedShardId, placementId,
+								SHARD_STATE_ACTIVE, ShardLength(colocatedShardId),
+								groupId);
+
+		if (ShouldSyncTableMetadata(colocatedShard->relationId))
+		{
+			char *placementCommand = PlacementUpsertCommand(colocatedShardId, placementId,
+															SHARD_STATE_ACTIVE, 0,
+															groupId);
+
+			SendCommandToWorkersWithMetadata(placementCommand);
+		}
+	}
+}
+
+
+/*
+ * EnsureTableListOwner ensures current user owns given tables. Superusers
+ * are regarded as owners.
+ */
+static void
+EnsureTableListOwner(List *tableIdList)
+{
+	Oid tableId = InvalidOid;
+	foreach_oid(tableId, tableIdList)
+	{
+		EnsureTableOwner(tableId);
+	}
+}
+
+
+/*
+ * EnsureTableListSuitableForReplication errors out if given tables are not
+ * suitable for replication.
+ */
+static void
+EnsureTableListSuitableForReplication(List *tableIdList)
+{
+	Oid tableId = InvalidOid;
+	foreach_oid(tableId, tableIdList)
+	{
+		char relationKind = get_rel_relkind(tableId);
+		if (relationKind == RELKIND_FOREIGN_TABLE)
+		{
+			char *relationName = get_rel_name(tableId);
+			ereport(ERROR, (errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+							errmsg("cannot replicate shard"),
+							errdetail("Table %s is a foreign table. Replicating "
+									  "shards backed by foreign tables is "
+									  "not supported.", relationName)));
+		}
+
+		List *foreignConstraintCommandList =
+			GetTableForeignConstraintCommands(tableId);
+
+		if (foreignConstraintCommandList != NIL &&
+			PartitionMethod(tableId) != DISTRIBUTE_BY_NONE)
+		{
+			ereport(ERROR, (errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
```

Comment:
We should probably add a test to cover this case.

Reply:
done